win32ss/user/ntuser/desktop.c

```diff
@@ -2957,26 +2958,36 @@ NtUserGetThreadDesktop(DWORD dwThreadId, DWORD Unknown1)
     }
 
     Status = PsLookupThreadByThreadId((HANDLE)(DWORD_PTR)dwThreadId, &Thread);
-    if (!NT_SUCCESS(Status))
+    if (!NT_SUCCESS(Status) || !Thread)
     {
-        EngSetLastError(ERROR_INVALID_PARAMETER);
-        RETURN(0);
+        Thread = NULL;
+        if (hConsoleDesktop)
+        {
+            hDesk = hConsoleDesktop;
+            process = gpepCSRSS;
+        }
+        else
+        {
+            EngSetLastError(ERROR_INVALID_PARAMETER);
+            RETURN(0);
+        }
+    }
+    else
+    {
+        hDesk = ((PTHREADINFO)Thread->Tcb.Win32Thread)->hdesk;
+        process = Thread->ThreadsProcess;
     }
 
-    if (Thread->ThreadsProcess == PsGetCurrentProcess())
+    if (process == PsGetCurrentProcess())
     {
         /* Just return the handle, we queried the desktop handle of a thread running
            in the same context */
-        hDesk = ((PTHREADINFO)Thread->Tcb.Win32Thread)->hdesk;
-        ObDereferenceObject(Thread);
         RETURN(hDesk);
     }
 
     /* Get the desktop handle and the desktop of the thread */
-    if (!(hThreadDesktop = ((PTHREADINFO)Thread->Tcb.Win32Thread)->hdesk) ||
-        !(DesktopObject = ((PTHREADINFO)Thread->Tcb.Win32Thread)->rpdesk))
```

Comment:
Why is it okay to remove the `rpdesk` check?

Reply:
Not exactly sure. This was one of the early patches, I don't remember the details. Now it feels like it's  a mistake. I will address many issues (especially C90 correctness) in the next few days, including this one. I also have a lot of other patches, but they were more tightly coupled and more risky standalone, these might take a week to prepare for PR.

Reply:
The goal of the function is to acquire the object, not do all possible checks and fail on some possible inconsistency (likely introduced by another component). The `DesktopObject` from `rpdesk` is not used (as explained below in a comment, safety reasons) and is always overwritten with newly referenced object, so no point in checking the cached value and failing even though we could complete the procedure successfully.